### PR TITLE
Update Parsing and Quoting articles

### LINKED ISSUE #2361: Document the need for additional \-escaping of double quotes (") when calling external programs
Related: #5152; once https://github.com/PowerShell/PowerShell/issues/1995 gets fixed, this topic will become obsolete.

An _additional_ layer of escaping of `"` chars. is unexpectedly needed when calling external programs from PowerShell to make sure that the external program sees `"` embedded in `"..."` as `\"`, which is what PowerShell's own CLI expects too.

This is currently not documented, as far I can tell, neither in [about_Parsing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing) nor in [about_Quoting_Rules](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules) nor in [about_Special_Characters](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_special_characters)

_Update_: As an alternative to the manual `\`-escaping detailed below, you can use the **PSv3+ `ie` helper function** from the  **[`Native` module](https://github.com/mklement0/Native)** (in PSv5+, install with `Install-Module Native` from the [PowerShell Gallery](https://www.powershellgallery.com/packages/Native)), which internally **compensates for all broken behavior** and allows passing arguments as expected; to use it, simply prepend `ie` to your invocations; e.g.:  
`ie pwsh -noprofile -command ' "hi" '` works as expected.

---

**Calling an external program _from PowerShell_** (the external program just happens to be another instance of PowerShell (`pwsh`) in the following examples):

In order to pass string literal `"hi"` as a command to an external program _from PowerShell_, **PowerShell's own escaping of embedded `"` chars. must unexpectedly be supplemented with `\`-escaping** (the following commands are equivalent):

```powershell
# !! The \-escaping of the " chars. should NOT be necessary; without it, these commands would fail.
PS> pwsh -noprofile -command " \""hi\"" " 
PS> pwsh -noprofile -command " \`"hi\`" "
PS> pwsh -noprofile -command ' \"hi\" '  
hi
```

<sup>As an aside: If you're really calling another PowerShell instance from inside PowerShell, using a _script block_ (`{ ... }`) to pass the command avoids all quoting headaches and additionally enables support for _typed_ results (not just strings), albeit with the same limitations on type fidelity as with background jobs / remoting.</sup>

Note: There are also a number of **bugs**:

* Empty-string arguments are quietly _removed_ from the invocation.

* Values with embedded `"` are passed incorrectly: in addition to not automatically escaping them (the problem shown above), the situationally necessary enclosure in `"..."` behind the scenes is not reliably triggered, such as with
`3" of snow`.

* In Windows PowerShell only, values with spaces that end in a `\` char. are passed incorrectly.

For details and workarounds, see https://github.com/PowerShell/PowerShell/issues/1995#issuecomment-562334606

---

**Calling the PowerShell CLI _from another shell_** - `cmd.exe` or `bash`:

Note: The difficulties discussed next stem primarily from `cmd.exe` limitations and how commands are invoked by a single command-line string on Window. 
Calling from Bash / POSIX-like shells works robustly and as expected.

---

`cmd.exe`:

On Windows, PowerShell _Core_ now properly recognizes as `""` as escaped `"`, which enables robust  escaping, given that `""` is also recognized by `cmd.exe` itself:

```sh
# PowerShell *Core* only, on Windows only: use "", which works robustly.
C:\> pwsh.exe -noprofile -command " ""hi   &   dry"" " 
hi  &  dry  # OK
```

Sadly, `""` doesn't work in _Windows PowerShell_, where `\`-escaping for PowerShell's sake is _also_ required, which causes problems:

Using `\""` (sic) doesn't require escaping of cmd.exe metachars., but doesn't preserve whitespace as-is:

```sh
# No extra escaping needed, but whitespace is normalized.
C:\> powershell.exe -noprofile -command " \""hi   &   dry\"" " 
hi & dry   # !! Two spaces were collapsed into one.
```

Only using `\"` preserves whitespace as-is, but it additionally _requires individual `^`-escaping of the following `cmd.exe` metacharacters_ inside `\"...\"` runs: `& | < > ^`

```sh
# Whitespace is faithfully preserved, but cmd.exe metachars. must be ^-escaped
C:\> powershell.exe -noprofile -command " \"hi   ^&   dry\" " 
hi  &  dry  # OK
```

---

`bash`:

Use `\"` inside `"...."` strings, and no escaping at all inside `'...'` strings:

```sh
$ pwsh -noprofile -command " \"hi\" " # \-escaping needed for Bash's own sake (nesting ")
$ pwsh -noprofile -command ' "hi" ' # !! NO escaping of " needed
```

Note that on Unix-like platforms PowerShell's own command-line parsing does _not_ come into play (arguments are invariably passed as an _array_ of _literal_ tokens), and the above commands _solely_ use Bash [non]-escaping to pass an argument with literal contents ` "hi" `.

Thus:

* `\"` is Bash's native way to escape `"` inside a `"..."` string.
* Inside `'...'`, `"` needs no escaping
  * Caveat: if you tried `\"` inside `'...'`, the `\` chars. would be passed as _literals_, and the PowerShell command would break.

This asymmetry with how things work on Windows is unfortunate, but unavoidable.

Note that calling _from PowerShell_ on Unix-like platforms _still_ requires the extra `\`-escaping:

```powershell
PS> bash -c 'echo "one two" '  # !! only prints 'one', because the " are *ignored*
PS> bash -c 'echo \"one two\" '  # OK, but the \-escaping shouldn't be necessary.
```

While this requirement at least makes for consistent behavior across platforms from the _PowerShell_ side, it is certainly unexpected - and an artificial requirement - for anyone familiar with Unix shell scripting.

---

The bottom line is:

* **In an ideal world, PowerShell would transparently handle any _additional_, platform-specific escaping needs for interfacing with external programs _behind the scenes_, so that all that users need to take care of - on any supported platform - is to follow _PowerShell_'s escaping rules.**

* Sadly, this is not an option if backward compatibility must be preserved.

Version(s) of document impacted
------------------------------
- [x] Impacts 6.1 document
- [x] Impacts 6.0 document
- [x] Impacts 5.1 document
- [x] Impacts 5.0 document
- [x] Impacts 4.0 document
- [x] Impacts 3.0 document

### LINKED ISSUE #3440: about_parsing additional expression mode start characters
Issue Details
-----------------

I have left the template blank at this time, but I believe this might affect all versions.

---
 - In argument mode, each value is treated as an expandable string unless it begins with one of the following special characters: dollar sign ($), at sign (@), single quotation mark ('), double quotation mark ("), or an opening parenthesis (().

If preceded by one of these characters, the value is treated as a value expression.

----

On about_parsing, there are additional characters that cause the expression mode, including:
- '-' or '+' or '!' as long as it is in turn followed by a numeric value or one of the other mentioned characters, ie, a valid expression.

`-$a` or `-2` or `!$a` is an expression mode
`-hello` or `+hello` or `!test` is an argument mode

Also, I think the last line should actually be part of the bulleted paragraph.

I think this whole section also fails to explain that the mode determination is based on the first token of each command statement and then applies to the remainder of the statement, noting that parenthesis and braces begin a new sub-statement.  Instead this section implies that the mode is determine for each token (it mentions the tokens are interpreted independently, but they are not, as interpretation of all tokens after the first one depend on the first one's evaluation).  Assuming a function named 'hello', `hello $a-3`, the token `$a-3` is still treated as an expandable string, not an expression.  An example in the doc also shows this, `Write-Output $a/H`.

Also, I think the word 'value' may have gotten used where 'token' should have been used, for consistency.  The doc started off talking about 'tokens' and then switches to 'values' in the bulleted items.

---
#### Document Details

 *Do not edit this section. It is required for docs.microsoft.com  GitHub issue linking.*

* ID: efd8c446-9ff3-8331-a996-4fe6226f2c6d
* Version Independent ID: b9957906-f58b-86de-7907-cc2f94174af3
* Content: [about_Parsing - PowerShell](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing?view=powershell-7)
* Content Source: [reference/7.0/Microsoft.PowerShell.Core/About/about_Parsing.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/live/reference/7.0/Microsoft.PowerShell.Core/About/about_Parsing.md)
* Product: **powershell**
* Technology: **powershell-conceptual**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

### LINKED ISSUE #6149: Document the specific constraints of the stop-parsing symbol (--%)
It is worth clarifying the following limitations, notably the one with respect to what causes `--%` to stop parsing (see https://github.com/PowerShell/PowerShell/issues/12975):

* `--%` supports only _one_ command, which an unquoted `|`, `||` or `&&` on the same line, if present, implicitly ends; this allows you to pipe / chain such a command to / with other commands.

  * However, using `;` in order to unconditionally place another command on the same line is _not_ supported; the `;` is passed through verbatim.

  *  `--%` reads (at most) to the _end of the line_ so spreading a command across multiple lines with line-continuation chars. is NOT supported

* Other than `%...%` *environment*-variable references, you cannot embed any other dynamic elements in the command; that is, **you cannot embed regular PowerShell variable references or expressions**.
   * Escaping `%` characters as `%%` (the way you can do inside batch files) is _not_  supported; `%<name>%` tokens are _invariably_ expanded, if `<name>` refers to a defined environment variable (if not, the token is passed through as-is). 

* You **cannot use stream redirections** (e.g., `>file.txt`), because they are passed _verbatim, as arguments_ to the target command.

  * For _stdout_ output you can work around that by appending `| Set-Content file.txt` instead, but there is no direct PowerShell workaround for _stderr_ output.

  * However, if you invoke your command _via `cmd`_, you can let `cmd` handle the (stderr) redirection (e.g., `cmd --% /c nosuch 2>file.txt`)

#4963 discusses why, specifically, `--%` is virtually useless on _Unix-like platforms_ (the `about_Parsing` topic now mentions that, but just in the abstract ("The stop-parsing symbol only intended for use on Windows platforms.")).

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b60e7291-d3a7-bf51-c945-ff5ef5f4d570
* Version Independent ID: b9957906-f58b-86de-7907-cc2f94174af3
* Content: [about_Parsing - PowerShell](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parsing?view=powershell-7.1)
* Content Source: [reference/7.1/Microsoft.PowerShell.Core/About/about_Parsing.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/7.1/Microsoft.PowerShell.Core/About/about_Parsing.md)
* Product: **powershell**
* GitHub Login: @SteveL-MSFT
* Microsoft Alias: **slee**

### PR DESCRIPTION
**Pull request by sdwheeler**
# PR Summary

Update Parsing and Quoting articles
- Fixes [AB#1873625](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1873625) - Fixes #6149
- Fixes [AB#1873626](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1873626) - Fixes #3440
- Fixes [AB#1873627](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1873627) - Fixes #2361

Select the area of the Table of Contents containing the documents being changed.

**Conceptual content**
- [ ] Overview and Install
- [ ] Learning PowerShell
  - [ ] PowerShell 101
  - [ ] Deep dives
  - [ ] Sample scripts
  - [ ] Remoting
- [ ] Release notes (What's New)
- [ ] Windows PowerShell
  - WMF, ISE, release notes, etc.
- [ ] DSC articles
- [ ] Community resources
- [ ] Gallery articles
- [ ] Scripting and development
  - [ ] Language Spec
  - [ ] Legacy SDK

**Cmdlet reference & about_ topics**
- [x] Preview content
- [x] Version 7.1 content
- [x] Version 7.0 content
- [x] Version 5.1 content

## PR Checklist

- [x] I have read the [contributors guide][contrib] and followed the style and process guidelines
- [x] PR has a meaningful title
- [x] PR is targeted at the _staging_ branch
- [x] All relevant versions updated
- [x] Includes content related to issues and PRs - see [Closing issues using keywords][key].
- [x] This PR is ready to merge and is not **Work in Progress**
  - If the PR is work in progress, please add the prefix `WIP:` or `[WIP]` to the beginning of the
    title and remove the prefix when the PR is ready.

[contrib]: https://docs.microsoft.com/powershell/scripting/community/contributing/overview
[key]: https://help.github.com/en/articles/closing-issues-using-keywords

### CODE REVIEW

File: reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md [hunk 1]
```diff
@@ -176,19 +200,80 @@ prevent PowerShell from misinterpreting the parentheses.
 icacls X:\VMS /grant Dom\HVAdmin:`(CI`)`(OI`)F
 ```
 
-Beginning in PowerShell 3.0, you can use the stop-parsing symbol.
+### The stop-parsing token
+
+Beginning in PowerShell 3.0, you can use the stop-parsing token (`--%`). The
+stop-parsing token directs PowerShell to refrain from interpreting input as
+PowerShell commands or expressions.
```
Comment (chasewilson) [inline reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md:207]:
```suggestion
Beginning in PowerShell 3.0, you can use the stop-parsing token (`--%`) to
stop PowerShell from interpreting input asPowerShell commands or expressions.
```

File: reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md [hunk 2]
```diff
@@ -74,96 +81,113 @@ Operators combine other expressions for evaluation:
...
+- `()`, `$()`, and `@()` at the start of a token create a new parsing context
+  that can contain expressions or nested commands.
+  - When followed by additional characters, the first additional character is
+    considered the start of a new, separate argument.
+  - When preceded by an unquoted literal `$()` works like an expandable string,
+    `()` starts a new argument that is an expression, and `@()` is taken as
+    literal `@` with `()` starting a new argument that is an expression.
+
+- Everything else is treated as an expandable string, except that
+  metacharacters still need escaping.
```
Comment (chasewilson) [inline reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md:129]:
```suggestion
- Everything else is treated as an expandable string, except
  metacharacters that still need escaping.
```

File: reference/7.0/Microsoft.PowerShell.Core/About/about_Parsing.md [hunk 3]
```diff
@@ -175,19 +200,70 @@ prevent PowerShell from misinterpreting the parentheses.
...
 
-Beginning in PowerShell 3.0, you can use the stop-parsing symbol.
+### The stop-parsing token
+
+Beginning in PowerShell 3.0, you can use the stop-parsing token (`--%`). The
+stop-parsing token directs PowerShell to refrain from interpreting input as
+PowerShell commands or expressions.
+
+> [!NOTE]
+> The stop-parsing token only intended for use on Windows platforms.
```
Comment (chasewilson) [inline reference/7.0/Microsoft.PowerShell.Core/About/about_Parsing.md:210]:
```suggestion
> The stop-parsing token is only intended for use on Windows platforms.
```

File: reference/7.1/Microsoft.PowerShell.Core/About/about_Parsing.md [hunk 4]
```diff
@@ -177,19 +200,70 @@ prevent PowerShell from misinterpreting the parentheses.
...
 
-Beginning in PowerShell 3.0, you can use the stop-parsing symbol.
+### The stop-parsing token
+
+Beginning in PowerShell 3.0, you can use the stop-parsing token (`--%`). The
+stop-parsing token directs PowerShell to refrain from interpreting input as
+PowerShell commands or expressions.
+
+> [!NOTE]
+> The stop-parsing token only intended for use on Windows platforms.
```
Comment (chasewilson) [inline reference/7.1/Microsoft.PowerShell.Core/About/about_Parsing.md:210]:
```suggestion
> The stop-parsing token is only intended for use on Windows platforms.
```

File: reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md [hunk 5]
```diff
@@ -176,19 +200,80 @@ prevent PowerShell from misinterpreting the parentheses.
...
 
-Beginning in PowerShell 3.0, you can use the stop-parsing symbol.
+### The stop-parsing token
+
+Beginning in PowerShell 3.0, you can use the stop-parsing token (`--%`). The
+stop-parsing token directs PowerShell to refrain from interpreting input as
+PowerShell commands or expressions.
+
+> [!NOTE]
+> The stop-parsing token only intended for use on Windows platforms.
```
Comment (chasewilson) [inline reference/7.2/Microsoft.PowerShell.Core/About/about_Parsing.md:210]:
```suggestion
> The stop-parsing token is only intended for use on Windows platforms.
```

File: reference/7.0/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 6]
```diff
@@ -59,53 +59,75 @@ The output of this command is:
...
+
+```powershell
+"PS version: $($PSVersionTable.PSVersion)"
+```
+
+```Output
+PS version: 7.0.7
+```
+
+To disambiguate a variable name from subsequent characters in the string,
```
Comment (chasewilson) [inline reference/7.0/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:74]:
```suggestion
To clarify a variable name from subsequent characters in the string,
```

File: reference/7.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 7]
```diff
@@ -59,53 +59,75 @@ The output of this command is:
...
+
+```powershell
+"PS version: $($PSVersionTable.PSVersion)"
+```
+
+```Output
+PS version: 7.1.4
+```
+
+To disambiguate a variable name from subsequent characters in the string,
```
Comment (chasewilson) [inline reference/7.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:74]:
```suggestion
To clarify a variable name from subsequent characters in the string,
```

File: reference/7.2/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 8]
```diff
@@ -59,53 +59,75 @@ The output of this command is:
...
+
+```powershell
+"PS version: $($PSVersionTable.PSVersion)"
+```
+
+```Output
+PS version: 7.2.0
+```
+
+To disambiguate a variable name from subsequent characters in the string,
```
Comment (chasewilson) [inline reference/7.2/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:74]:
```suggestion
To clarify a variable name from subsequent characters in the string,
```

File: reference/7.0/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 9]
```diff
@@ -308,6 +333,43 @@ Here-strings are also a convenient format for input to the
 `ConvertFrom-StringData` cmdlet, which converts here-strings to hash tables.
 For more information, see `ConvertFrom-StringData`.
 
+## Interpretation of expandable strings
+
+Expanded string don't necessarily look the same as the default output that you
```
Comment (chasewilson) [inline reference/7.0/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:338]:
```suggestion
Expanded strings don't necessarily look the same as the default output that you
```

File: reference/7.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 10]
```diff
@@ -308,6 +333,43 @@ Here-strings are also a convenient format for input to the
 `ConvertFrom-StringData` cmdlet, which converts here-strings to hash tables.
 For more information, see `ConvertFrom-StringData`.
 
+## Interpretation of expandable strings
+
+Expanded string don't necessarily look the same as the default output that you
```
Comment (chasewilson) [inline reference/7.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:338]:
```suggestion
Expanded strings don't necessarily look the same as the default output that you
```

File: reference/7.2/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 11]
```diff
@@ -308,6 +333,43 @@ Here-strings are also a convenient format for input to the
 `ConvertFrom-StringData` cmdlet, which converts here-strings to hash tables.
 For more information, see `ConvertFrom-StringData`.
 
+## Interpretation of expandable strings
+
+Expanded string don't necessarily look the same as the default output that you
```
Comment (chasewilson) [inline reference/7.2/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:338]:
```suggestion
Expanded strings don't necessarily look the same as the default output that you
```

**Review by chasewilson — CHANGES_REQUESTED** [review]

File: reference/5.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md [hunk 12]
```diff
@@ -59,53 +59,75 @@ The output of this command is:
...
+
+```powershell
+"PS version: $($PSVersionTable.PSVersion)"
+```
+
+```Output
+PS version: 5.1.19041.1151
+```
+
+To disambiguate a variable name from subsequent characters in the string,
```
[the remark this answers is not in the document]
Comment (sdwheeler) [reply reference/5.1/Microsoft.PowerShell.Core/About/about_Quoting_Rules.md:74]: I am not crazy about 'disabiguate' but 'clarify' is not the same thing.

**Review by chasewilson — APPROVED** [review]

**Pull request merged**